// View/Music/index.js
/**
 * Created by leewoongjae on 15. 7. 6..
 */
module.exports = (function() {
    "use strict";

    var $ = global.window.$;
    var document = global.window.document; // audio 찾기 위함

    var application = global.getApplication();

    var musicList = [

        {
            "rank"	: "1",
            "title"	: "A Day in the Life",
            "subtitle"	: "Sgt. Pepper's Lonely Hearts Club Band(1967)",
            "audio"	: "01_adayinthelife",
            "background"	: "01_adayinthelife"
        },
        {
            "rank"	: "2",
            "title"	: "Strawberry Fields Forever",
            "subtitle"	: "1967 single",
            "audio"	: "02_strawberryfieldsforever",
            "background"	: "02_strawberryfieldsforever"
        },
        {
            "rank"	: "3",
            "title"	: "Abbey Road Medley",
            "subtitle"	: "Abbey Road(1969)",
            "audio"	: "03_abbeyroadmedley",
            "background"	: "03_abbeyroadmedley"
        },
        {
            "rank"	: "4",
            "title"	: "While My Guitar Gently Weeps",
            "subtitle"	: "The Beatles(1968)",
            "audio"	: "04_whilemyguitargentlyweeps",
            "background"	: "04_whilemyguitargentlyweeps"
        },
        {
            "rank"	: "5",
            "title"	: "In My Life",
            "subtitle"	: "Rubber Soul(1965)",
            "audio"	: "05_inmylife",
            "background"	: "05_inmylife"
        },
        {
            "rank"	: "6",
            "title"	: "Let It Be",
            "subtitle"	: "1970 single",
            "audio"	: "06_letitbe",
            "background"	: "06_letitbe"
        },
        {
            "rank"	: "7",
            "title"	: "Something",
            "subtitle"	: "Abbey Road(1969)",
            "audio"	: "07_something",
            "background"	: "07_something"
        },
        {
            "rank"	: "8",
            "title"	: "Penny Lane",
            "subtitle"	: "1967 single",
            "audio"	: "08_pennylane",
            "background"	: "08_pennylane"
        },
        {
            "rank"	: "9",
            "title"	: "I Want to Hold Your Hand",
            "subtitle"	: "1963 single",
            "audio"	: "09_iwanttoholdyourhand",
            "background"	: "09_iwanttoholdyourhand"
        },
        {
            "rank"	: "10",
            "title"	: "Ticket to Ride",
            "subtitle"	: "Help!(1965)",
            "audio"	: "10_tickettoride",
            "background"	: "10_tickettoride"
        },
        {
            "rank"	: "11",
            "title"	: "Hey Jude",
            "subtitle"	: "1968 Single",
            "audio"	: "11_heyjude",
            "background"	: "11_heyjude"
        },
        {
            "rank"	: "12",
            "title"	: "Yesterday",
            "subtitle"	: "Help!(1965)",
            "audio"	: "12_yesterday",
            "background"	: "12_yesterday"
        },
        {
            "rank"	: "13",
            "title"	: "Help!",
            "subtitle"	: "1965 Single",
            "audio"	: "13_help",
            "background"	: "13_help"
        },
        {
            "rank"	: "14",
            "title"	: "She Loves You",
            "subtitle"	: "1963 Single",
            "audio"	: "14_shelovesyou",
            "background"	: "14_shelovesyou"
        },
        {
            "rank"	: "15",
            "title"	: "Norwegian Wood",
            "subtitle"	: "Rubber Soul(1965)",
            "audio"	: "15_norwegianwood",
            "background"	: "15_norwegianwood"
        },
        {
            "rank"	: "16",
            "title"	: "Eleanor Rigby",
            "subtitle"	: "1966 Single",
            "audio"	: "16_eleanorrigby",
            "background"	: "16_eleanorrigby"
        },
        {
            "rank"	: "17",
            "title"	: "I Saw Her Standing There",
            "subtitle"	: "1963 Single",
            "audio"	: "17_isawherstandingthere",
            "background"	: "17_isawherstandingthere"
        },
        {
            "rank"	: "18",
            "title"	: "A Hard Day's Night",
            "subtitle"	: "A Hard Day's Night(1964)",
            "audio"	: "18_aharddaysnight",
            "background"	: "18_aharddaysnight"
        },
        {
            "rank"	: "19",
            "title"	: "Please Please Me",
            "subtitle"	: "1963 Single",
            "audio"	: "19_pleasepleaseme",
            "background"	: "19_pleasepleaseme"
        },
        {
            "rank"	: "20",
            "title"	: "Come Together",
            "subtitle"	: "Abbey Road(1969)",
            "audio"	: "20_cometogether",
            "background"	: "20_cometogether"
        }

    ];


    var MusicPlayer = function() {

        this.dom = $("#Scene-MusicPlayer");

        this.make();

        this.audio = $("#audio");

        this.isPlay = false;
        this.index = 0;

        this.update = setInterval(function() {

            if($("#audio")[0].currentTime){

                var strTime = asTime($("#audio")[0].currentTime);
                var strDuration = asTime($("#audio")[0].duration);
                $("#currentTime").text(strTime);
                $("#totalTime").text(strDuration);

                var percent = $("#audio")[0].currentTime / $("#audio")[0].duration;

                var length = percent * parseFloat($("#noti-total-time").css("width"));
                $("#noti-current-time").css("width", length);

            }

        }, 100);

        this.playMusic();

    };

    var asTime = function(t) {

        t = Math.round(t);
        var s = t % 60;
        var m = Math.floor(t / 60);
        return two(m) + ':' + two(s);

    };

    var two = function(s) {
        s += "";
        if (s.length < 2) s = "0" + s;
        return s;
    }

    MusicPlayer.prototype.make = function () {

        var self = this;

        self.dom.append("<audio id=\"select\" src=\"audio/select.mp3\"></audio>" +
                        "<div id=\"prev\" class=\"button\" >" +
                            "<img src=\"images/left.png\"/>" +
                        "</div>" +
                        "<div id=\"bg_play\">" +
                            "<div id=\"play\" class=\"button\">" +
                                "<img id=\"icon_play\" src=\"images/play.png\"/>" +
                                "<div class=\"border\"></div>" +
                            "</div>" +
                        "</div>" +
                        "<div id=\"next\" class=\"button\">" +
                            "<img src=\"images/right.png\"/>" +
                        "</div>" +
                        "<ul id=\"background\"></ul>" +
                        "<div id=\"time\">" +
                            "<p id=\"currentTime\" align=\"center\"> 00:00 </p>" +
                            "<div id=\"noti-total-time\">" +
                                "<div id=\"noti-current-time\"></div>" +
                            "</div>" +
                            "<p id=\"totalTime\" align=\"center\"> 00:00 </p>" +
                        "</div>" +
                        "<div id=\"logo\">" +
                            "<img src=\"images/logo.png\"/>" +
                        "</div>" +
                        "<audio id=\"audio\" >" +
                            "<source id=\"mp3_src\" src=\"audio/01_adayinthelife.mp3\" type=\"audio/mp3\"></source>" +
                        "</audio>" +
                        "<div id=\"turntable\">" +
                            "<div class=\"border\"></div>" +
                        "</div>" +
                        "<div id=\"title-bar\"></div>" +
                        "<div id = \"hand_cursor\"></div>");

        for (var i = 0; i < musicList.length; i++) {

            $('#background').append("<li>" +
                                        "<div class=\"cover\">" +
                                            "<img style=\"width:800px;left:0px;\" src=\"images/" + musicList[i].background + ".jpg\"/>" +
                                        "</div>" +
                                        "<div id=\"title\" align=\"center\">" + musicList[i].rank + ". " + musicList[i].title + "</div>" +
                                        "<div id=\"subtitle\" align=\"center\">" + musicList[i].subtitle + "</div>" +
                                    "</li>");

        }

        $('#background li').eq(19).addClass("left");
        $('#background li').eq(0).addClass("center");
        $('#background li').eq(1).addClass("right");

    };

    MusicPlayer.prototype.show = function () {

        this.dom.fadeIn();

    };

    MusicPlayer.prototype.hide = function () {

        this.dom.fadeOut();

    };

    MusicPlayer.prototype.trigger = function (vtouch, isLOCKHIT, isLOCKAREA, isDISPLAY, isHUE1, isHUE2, isPHOTO, isTHERMO, isMUSIC) {

        var self = this;

        var Users = application.getUsers();

        if (self.isFirstUser > -1) {

            if (Users[self.isFirstUser].vtouch === undefined) return;
            if (Users[self.isFirstUser].vtouch === null) return;

            if (vtouch[self.isFirstUser].trigger == "U") {

                var touch = Users[self.isFirstUser].vtouch;

                if (touch.id == "R") self.right("U");
                else if (touch.id == "L") self.left("U");
                else if (touch.id == "T") self.up("U");
                else if (touch.id == "B") self.down("U");
                else if (touch.id == "DISPLAY") self.display("U");

                self.isFirstUser = -1;

            } else if (vtouch[self.isFirstUser].trigger == "NU") {

                self.isFirstUser = -1;

            } else if (vtouch[self.isFirstUser].trigger == "N") {

                self.isFirstUser = -1;

            } else if (vtouch[self.isFirstUser].trigger == "D") {

                var touch = Users[self.isFirstUser].vtouch;

                if (touch.id == "R") self.right("D");
                else if (touch.id == "L") self.left("D");
                else if (touch.id == "T") self.up("D");
                else if (touch.id == "B") self.down("D");
                else if (touch.id == "DISPLAY") self.display("D");

            }

        } else {

            if (isDISPLAY > -1) {

                var isFirst = -1;

                for (var i = 0; i < 6; i++) {

                    if (Users[i].tracking == 1 && Users[i].triggerState == 0 && Users[i].isDownCount == 1) {

                        var touch = (Users[i].isRight) ? vtouch[i].right : vtouch[i].left;

                        if (touch.isHit) {

                            if (touch.id == "DISPLAY" || touch.id == "T" || touch.id == "B" || touch.id == "L" || touch.id == "R") isFirst = i;

                        }

                    }

                }

                if (isFirst > -1) {

                    self.isFirstUser = isFirst;

                    var touch = (Users[self.isFirstUser].isRight) ? vtouch[self.isFirstUser].right : vtouch[self.isFirstUser].left;

                    if (touch.isHit) {

                        Users[self.isFirstUser].vtouch = {};
                        Users[self.isFirstUser].vtouch.id = touch.id;
                        Users[self.isFirstUser].vtouch.point = {};
                        Users[self.isFirstUser].vtouch.point.x = touch.point.x;
                        Users[self.isFirstUser].vtouch.point.y = touch.point.y;

                        if (touch.id == "R") self.right("D");
                        else if (touch.id == "L") self.left("D");
                        else if (touch.id == "T") self.up("D");
                        else if (touch.id == "B") self.down("D");
                        else if (touch.id == "DISPLAY") self.display("D");

                    }

                }

            }

        }

        if (isLOCKAREA > -1) application.trigger_unlock(vtouch, isLOCKAREA);

    };

    MusicPlayer.prototype.left = function (_trigger) {

        var self = this;

        beepSound();

        self.index = (self.index > 0) ? (self.index - 1) : musicList.length - 1;
        $(".active").removeClass("active");

        self.changeMusic(musicList[self.index].rank, musicList[self.index].title, musicList[self.index].subtitle, musicList[self.index].audio, 1);

    };

    MusicPlayer.prototype.right = function (_trigger) {

        var self = this;

        beepSound();

        self.index = (self.index < musicList.length - 1) ? (self.index + 1) : 0;

        self.changeMusic(musicList[self.index].rank, musicList[self.index].title, musicList[self.index].subtitle, musicList[self.index].audio, 0);

    };

    MusicPlayer.prototype.changeMusic = function(_rank, _title, _subTitle, _music, _direction){

        var self = this;

        if(_direction === 0) // right
        {

            $( ".left" ).hide( "slide", {direction:'left'}, 500, function(){

                $(this).removeClass("left");

            })

            $( ".center" ).switchClass( "center", "left", 500 );

            $( ".right" ).switchClass( "right", "center", 500 , function(){

                $(".center").addClass("active");

            });

            if(_rank == 20) $('#background li').eq(0).addClass("right").hide().show("slide", {direction:'right'}, 500);
            else $( ".right").next().addClass("right").hide().show("slide", {direction:'right'}, 500);

        } else if (_direction === 1 ) { // left

            $( ".right" ).hide( "slide", {direction:'right'}, 500, function(){

                $(this).removeClass("right");

            });
            $( ".center" ).switchClass( "center", "right", 500 );

            $( ".left" ).switchClass( "left", "center", 500, function(){

                $(".center").addClass("center");

            });

            if(_rank == 1) $('#background li').eq(19).addClass("left").hide().show("slide", {direction:'left'}, 500);
            else $( ".left").prev().addClass("left").hide().show("slide", {direction:'left'}, 500);

        }

        $("#mp3_src").attr("src", "audio/" + _music + ".mp3" );

        self.audio[0].pause();
        self.audio[0].load();

        self.playMusic();

    };

    MusicPlayer.prototype.top = function (_trigger) {

        var self = this;

    };

    MusicPlayer.prototype.bottom = function (_trigger) {

        var self = this;

    };

    MusicPlayer.prototype.display = function (_trigger) {

        var self = this;

    };

    MusicPlayer.prototype.toggle = function () {

        var self = this;

        beepSound();

        if (self.audio[0].paused) self.playMusic();
        else self.pauseMusic();

    };

    MusicPlayer.prototype.playMusic = function() {

        this.audio[0].play();
        $("#play img").attr("src", "images/stop.png");

    };

    MusicPlayer.prototype.pauseMusic = function() {

        this.audio[0].pause();
        $("#play img").attr("src", "images/play.png");

    };

    var beepSound = function() {

        var beep = document.getElementById("select");
        if(beep !== null) beep.play();

    };

    return MusicPlayer;

}());
